server/core_notification.go

```diff
@@ -0,0 +1,215 @@
+			logger.Error("Could not scan notification from database.", zap.Error(err))
+			return nil, err
+		}
+		notifications = append(notifications, no)
+	}
+
+	notificationList := &api.NotificationList{}
+	cursorBuf := new(bytes.Buffer)
+	if len(notifications) == 0 {
+		if len(cursor) > 0 {
+			notificationList.CacheableCursor = cursor
+		} else {
+			newCursor := &notificationCacheableCursor{NotificationID: ""}
+			if err := gob.NewEncoder(cursorBuf).Encode(newCursor); err != nil {
+				logger.Error("Could not create new cursor.", zap.Error(err))
+				return nil, err
+			}
+			notificationList.CacheableCursor = base64.RawURLEncoding.EncodeToString(cursorBuf.Bytes())
+		}
+	} else {
+		lastNotification := notifications[len(notifications)-1]
+		newCursor := &notificationCacheableCursor{
+			NotificationID: lastNotification.Id,
+		}
+		if err := gob.NewEncoder(cursorBuf).Encode(newCursor); err != nil {
+			logger.Error("Could not create new cursor.", zap.Error(err))
+			return nil, err
+		}
+		notificationList.Notifications = notifications
+		notificationList.CacheableCursor = base64.RawURLEncoding.EncodeToString(cursorBuf.Bytes())
+	}
+
+	return notificationList, nil
+}
+
+func NotificationDelete(logger *zap.Logger, db *sql.DB, userID uuid.UUID, notificationIDs []string) error {
+	statements := make([]string, 0, len(notificationIDs))
+	params := make([]interface{}, 0, len(notificationIDs))
+	params = append(params, userID)
+
+	for _, id := range notificationIDs {
+		statement := "$" + strconv.Itoa(len(params)+1)
+		statements = append(statements, statement)
+		params = append(params, id)
+	}
+
+	_, err := db.Exec("DELETE FROM notification WHERE user_id = $1 AND id IN ("+strings.Join(statements, ", ")+")", params...)
+	if err != nil {
+		logger.Error("Could not delete notifications.", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func NotificationSave(logger *zap.Logger, db *sql.DB, notifications map[uuid.UUID][]*api.Notification) error {
+	statements := make([]string, 0, len(notifications))
+	params := make([]interface{}, 0, len(notifications))
+	counter := 0
+	for userID, no := range notifications {
+		for _, un := range no {
+			statement := "$" + strconv.Itoa(counter+1) +
+				",$" + strconv.Itoa(counter+2) +
+				",$" + strconv.Itoa(counter+3) +
+				",$" + strconv.Itoa(counter+4) +
+				",$" + strconv.Itoa(counter+5) +
+				",$" + strconv.Itoa(counter+6)
+
+			if un.SenderId == "" {
+				statement += ",NULL"
+			} else {
+				statement += ",$" + strconv.Itoa(counter+7)
+			}
+
+			statements = append(statements, "("+statement+")")
+
+			params = append(params, un.Id)
+			params = append(params, userID)
+			params = append(params, un.Subject)
+			params = append(params, un.Content)
+			params = append(params, un.Code)
+			params = append(params, un.CreateTime.Seconds)
+			if un.SenderId != "" {
+				params = append(params, un.SenderId)
+			}
+
+			counter = counter + 8
```

Comment:
Is this accurate? Looks to me like the counter should advance by 6 in most cases, or by 7 when there is a sender ID.

Reply:
Good catch.